controller/konnect/reconciler_credential_secrets.go

```diff
@@ -0,0 +1,418 @@
+			return res, nil
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+const (
+	// CredentialTypeLabel is the label key for the credential type.
+	CredentialTypeLabel = "konghq.com/credential" //nolint:gosec
+)
+
+func ensureExistingCredential[
+	T constraints.SupportedCredentialType,
+	TPtr constraints.KongCredential[T],
+](
+	ctx context.Context,
+	cl client.Client,
+	cred TPtr,
+	secret *corev1.Secret,
+	consumer *configurationv1.KongConsumer,
+) (ctrl.Result, error) {
+	var update bool
+
+	switch cred := any(cred).(type) {
+	case *configurationv1alpha1.KongCredentialBasicAuth:
+		if cred.Spec.ConsumerRef.Name != consumer.Name ||
+			cred.Spec.Password != string(secret.Data[corev1.BasicAuthPasswordKey]) ||
+			cred.Spec.Username != string(secret.Data[corev1.BasicAuthUsernameKey]) {
+
+			cred.Spec.ConsumerRef.Name = consumer.Name
+			setKongCredentialBasicAuthSpec(&cred.Spec, secret)
+			update = true
+		}
+
+	default:
+		// NOTE: Shouldn't happen.
+		panic(fmt.Sprintf("unsupported credential type %T", cred))
+	}
+
+	if update {
+		if err := cl.Update(ctx, cred); err != nil {
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{
+					Requeue: true,
+				}, nil
+			}
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// ensureCredentialExists creates the credential based on the provided secret
+// and credType.
+func (r *KongCredentialSecretReconciler) ensureCredentialExists(
+	ctx context.Context,
+	secret *corev1.Secret,
+	credType string,
+	kongConsumer *configurationv1.KongConsumer,
+) error {
+	nn := types.NamespacedName{
+		Namespace: secret.Namespace,
+		Name:      secret.Name,
+	}
+
+	var cred client.Object
+	switch credType {
+	case KongCredentialTypeBasicAuth:
+		cred = secretToKongCredentialBasicAuth(secret, kongConsumer)
+
+	default:
+		return fmt.Errorf("Secret %s used as credential, but has unsupported type %s",
+			nn, credType,
+		)
+	}
+
+	err := controllerutil.SetControllerReference(kongConsumer, cred, r.scheme)
+	if err != nil {
+		return err
+	}
+	// Set the secret as owner too so that deletion (or changes) of the credential
+	// triggers the reconciliation of the secret in this controller.
+	err = controllerutil.SetOwnerReference(secret, cred, r.scheme)
+	if err != nil {
+		return err
+	}
+	if err := r.client.Create(ctx, cred); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// extractKongCredentialType returns the credential type of a Secret extracted from
+// the konghq.com/credential label.
+// If the label is not present it returns an error.
+func extractKongCredentialType(secret *corev1.Secret) (string, error) {
+	credType, ok := secret.Labels[CredentialTypeLabel]
+	if !ok {
+		return "", fmt.Errorf("Secret %s used as credential, but lacks %s label",
+			client.ObjectKeyFromObject(secret), CredentialTypeLabel,
+		)
+	}
+	return credType, nil
+}
+
+func validateSecret(
+	s *corev1.Secret,
+	credType string,
+) error {
+	nn := client.ObjectKeyFromObject(s)
+
+	switch credType {
```

Comment:
should we add some note here about the fact that additional credential types will be added in the future?

Reply:
👍 4c3f5a9411943b1dcef2f2e161305538ed312e53